File: Algorithms/p101_Symmetric_Tree_3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(!root) return 1;
        queue<pair<TreeNode*,TreeNode*>> q;
        q.emplace(root->left,root->right);
        while(!q.empty()){
            pair<TreeNode*,TreeNode*> node = q.front();
            q.pop();
            if(!node.first&&!node.second) continue;
            if(!node.first||!node.second) return 0;
            if(node.first->val!=node.second->val) return 0;
            q.emplace(node.first->left,node.second->right);
            q.emplace(node.first->right,node.second->left);
        }
        return 1;
    }
};

File: Algorithms/p101_Symmetric_Tree/p101_Symmetric_Tree_3.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isSymmetric(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if not root:
            return True
        q = collections.deque([(root.left,root.right)])
        while q:
            node = q.popleft()
            if not node[0] and not node[1]:
                continue
            if not node[0] or not node[1]:
                return False
            if node[0].val!=node[1].val:
                return False
            q.append((node[0].left,node[1].right))
            q.append((node[0].right,node[1].left))
        return True
